# Assignment2 - Ning

#### 1.Steps for solving the problem：

* understand question and test case
* check input
* algorithm
* check return

#### 2.Instruction: 
* make sure your code pass the assertion statements


### Q1. Given a positive integer N. The task is to write a Python program to check if the number is prime or not.

In [173]:
def is_prime(n: int) -> bool:
    if n < 0 or int(n) != n:
        print(n, "不是自然数，请输入一个自然数")
        return False
    elif n == 0 or n == 1:
        print(n, "既不是质数，也不是合数")
        return False
    elif n == 2:
        return True
    else:
        for i in range(2, n):
            if n%i == 0:
                return False
                break      # 在此处break可以在n为合数时省去几次循环
        else:
            return True

In [61]:
# DO NOT ALTER BELOW.
assert is_prime(2)
assert not is_prime(15)
assert is_prime(7907)
assert not is_prime(-1)
assert not is_prime(0)

-1 不是自然数，请输入一个自然数
0 既不是质数，也不是合数


### Q2 Write a function rotate(ar[], d) that rotates arr[] of size n by d elements.
* Input ar = [1,2,3,4,5,6,7], d = 2
* Output [3,4,5,6,7,1,2]

In [82]:
def rotate(ar: [int], d: int) -> [int]:
    for i in range(d):
        p = ar[0]
        ar.append(p)
        ar.pop(0)
    return ar

In [83]:
# DO NOT ALTER BELOW.
assert rotate([1,2,3,4,5,6,7], 2) == [3,4,5,6,7,1,2]
assert rotate([1,2,3], 4) == [2,3,1]

### Q3 Implement a workable selection sort algorithm (Selection sort)
* https://www.runoob.com/w3cnote/selection-sort.html 作为参考
* Input students would be a list of [student #, score], sort by score ascending order.

In [26]:
def selection_sort(arr: [[int]]) -> [[int]]:
    score = list()
    score_s = list()
    for i in range(len(arr)):
        score.append(arr[i][1])
    score.sort()
    for j in range(len(arr)):
        for p in range(len(arr)):
            if arr[p][1] == score[j]:
                score_s.append(arr[p])
    return score_s

In [28]:
# DO NOT ALTER BELOW.
assert selection_sort([]) == []
assert selection_sort([[1, 100], [2, 70], [3, 95], [4, 66], [5, 98]]) == [[4, 66], [2, 70], [3, 95], [5, 98], [1, 100]]

### Q4. Convert a list of Tuples into Dictionary
* tips: copy operation - copy by value, copy by reference
* Do NOT RETURN di, EDIT IN-PLACE

In [167]:
expected_dict = {}
def convert(tup: (any), di: {any, any}) -> None: 
    for i in range(0, len(tup), 2):
        di[tup[i]] = tup[i+1]
    return

In [168]:
# DO NOT ALTER BELOW.
expected_dict = {}
convert((), expected_dict)
assert expected_dict == {}

convert(('key1', 'val1', 'key2', 'val2'), expected_dict)
assert expected_dict == {'key1': 'val1', 'key2': 'val2'} 

### Q5. Find left-most and right-most index for a target in a sorted array with duplicated items.
* provided an example of slow version of bsearch_slow with O(n) time complecity. 
* your solution should be faster than bsearch_slow

In [169]:
import timeit
def create_arr(count: int, dup: int) -> [int]:
    return [dup for i in range(count)]

In [170]:
def bsearch_slow(arr: [int], target: int) -> (int):
    left = -1
    right = -1
    for i in range(len(arr)):
        if arr[i] == target and left == -1:
            left = i
        if arr[i] > target and left != -1 and right == -1:
            right = i
        if i == len(arr) - 1:
            right = len(arr) - 1
    return (left, right)

In [107]:
assert bsearch_slow(create_arr(10000, 5), 5) == (0, 9999)

In [ ]:
timeit.timeit(lambda: bsearch_slow(create_arr(10000, 5), 5), number=100)

* slow version runing 100 times = 0.39607857299961324 seconds

### Add your version and compare if faster.

In [103]:
# Complete this 
def bsearch(arr: [int], target: int) -> int:
    # 先找最左边的目标值
    l1 = 0
    r1 = len(arr)-1
    while(l1 < r1):
        mid = (l1 + r1)//2
        if(arr[mid] < target):
            l1 = mid+1
        else:
            r1 = mid
    if(arr[r1] == target):
        left = r1
    # 再找最右边的目标值
    l2 = 0
    r2 = len(arr)-1        
    if arr[r2] == target:
        right = r2
    else:
        while(l2 < r2):
            mid = (l2 + r2)//2
            if(arr[mid] <= target):
                l2 = mid 
            else:
                r2 = mid-1
        if(arr[l2] == target):
            right = l2
    return(left, right)

(0, 999)


In [109]:
assert bsearch(create_arr(1000, 5), 5) == (0, 999)

In [105]:
timeit.timeit(lambda: bsearch(create_arr(10000, 5), 5), number=100)

0.05680650500107731

* fast version runing 100 times = 0.05680650500107731 seconds
* 比slow version要快，slow version的时间复杂度为O(N)，空间复杂度也为O(1)；而利用二分查找空间复杂度为O(1)，时间复杂度为O(2logN)

### Q6. 为什么 Python dict 可以做到常数级别的查找效率

#### 1.详解Python字典的底层原理——哈希表(Python面试必备)
https://blog.csdn.net/qq_42815145/article/details/91353624?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522162866089316780262546656%2522%252C%2522scm%2522%253A%252220140713.130102334.pc%255Fall.%2522%257D&request_id=162866089316780262546656&biz_id=0&utm_medium=distribute.pc_search_result.none-task-blog-2~all~first_rank_v2~rank_v29-5-91353624.pc_search_download_positive&utm_term=Python+dict+hash+%E7%BA%A2%E9%BB%91%E6%A0%91&spm=1018.2226.3001.4187
#### 2.python字典实现原理-哈希函数-解决哈希冲突方法
https://blog.csdn.net/weixin_34293141/article/details/94218306?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522162866089316780262546656%2522%252C%2522scm%2522%253A%252220140713.130102334.pc%255Fall.%2522%257D&request_id=162866089316780262546656&biz_id=0&utm_medium=distribute.pc_search_result.none-task-blog-2~all~first_rank_v2~rank_v29-11-94218306.pc_search_download_positive&utm_term=Python+dict+hash+%E7%BA%A2%E9%BB%91%E6%A0%91&spm=1018.2226.3001.4187